### ISSUE TITLE
[MPS] torch.fft.fftn SIGABRT on high-rank complex64 inputs with explicit dim

### ISSUE THREAD

**Issue by christianahmed905** [pytorch/pytorch#186885]
### 🐛 Describe the bug

"""
torch.fft.fftn — SIGABRT (process crash) on MPS for high-rank complex inputs

torch.fft.fftn crashes the Python process with SIGABRT (exit code -6) on MPS
when the input tensor has ndim > 4 and a multi-axis dim tuple is supplied.
No Python exception is raised; the process is killed and cannot be recovered.

Crash origin:
  MPSGraphExecutable.mm:1484: failed assertion `original module failed verification'

Root cause: the MPS FFT path builds an MPSGraph executable for the requested
transform axes.  For 6D complex64 inputs with dim=(1, 2), graph verification
fails inside MetalPerformanceShadersGraph and triggers a fatal assertion.
PyTorch does not validate this constraint before dispatching to MPS.

Crash statistics (Specialized generator, mps):
  torch.fft.fftn — 1 crash (hash [PROC], signal=6)
                   example: shape=(2, 1, 1, 1, 2, 2) rank=6 dim=(1, 2)
                            norm=forward dtype=torch.complex64

Expected: transformed tensor with the same shape, OR a catchable RuntimeError
          explaining unsupported rank/axis combinations on MPS.
Actual:   SIGABRT kills the Python process — not catchable with try/except.

Affected op:     torch.fft.fftn (and likely related fftn/irfftn variants)
Affected device: mps (Apple Silicon)
Not affected:    same inputs on CPU; lower-rank MPS inputs (ndim <= 4)
Versions:        PyTorch 2.12.0, macOS 26.4.1 arm64 (Apple Silicon MPS)
"""

import torch

print(f"torch version: {torch.__version__}")
assert torch.backends.mps.is_available(), "MPS not available — run on Apple Silicon"

device = torch.device("mps")

# ── Safe inputs (ndim <= 4) ─────────────────────────────────────────────────
x4 = torch.randn(2, 4, 8, 8, dtype=torch.complex64, device=device)
print(f"[ok] fftn ndim=4 dim=(1, 2): {torch.fft.fftn(x4, dim=(1, 2), norm='forward').shape}")

# ── Safe high-rank input (transform all axes) ─────────────────────────────────
x6_all = torch.randn(2, 1, 1, 1, 2, 2, dtype=torch.complex64, device=device)
print(f"[ok] fftn ndim=6 dim=None:   {torch.fft.fftn(x6_all, dim=None, norm='forward').shape}")

print()
print("Choosing which crash to trigger (pick one — each kills the process):")
print("  Uncomment exactly one block below and re-run.\n")

# WARNING: the block below kills the process with SIGABRT.
# Uncomment to reproduce the fuzz crash.

# -- fftn ndim=6 dim=(1, 2) complex64 (fuzz example) --
# x6 = torch.randn(2, 1, 1, 1, 2, 2, dtype=torch.complex64, device=device)
# print(f"input shape: {tuple(x6.shape)}  dim=(1, 2)  norm=forward  dtype={x6.dtype}")
# print("torch.fft.fftn(x6, dim=(1, 2), norm='forward') → SIGABRT")
# torch.fft.fftn(x6, dim=(1, 2), norm='forward')

print("(no crash triggered — uncomment a block above to reproduce)")

print("\n[ref] CPU (same inputs, should succeed):")
x6_cpu = torch.randn(2, 1, 1, 1, 2, 2, dtype=torch.complex64)
print(f"  fftn ndim=6 dim=(1, 2): {torch.fft.fftn(x6_cpu, dim=(1, 2), norm='forward').shape}")

### Versions

Collecting environment information...
PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.5 (main, May 10 2026, 19:20:57) [Clang 22.1.3 ] (64-bit runtime)
Python platform: macOS-26.5-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M5

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[conda] Could not collect

cc @ezyang @anjali411 @dylanbespalko @mruberry @nikitaved @amjames @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

**Comment by adhi29** [comment]
Can I work on this issue?

**Comment by jhavukainen** [comment]
Issue is fixed by https://github.com/pytorch/pytorch/pull/186967: The MPSGraph function supports the transformation only in the last four indices of the tensor. Implementation now permutes the tensor if necessary and raises an error if it can't be done.
